src/Tableau.cpp

```diff
@@ -0,0 +1,395 @@
+                tableau[i][2U * nqubits] ^= tableau[i][a] & tableau[i][a + nqubits];
+                tableau[i][a + nqubits] ^= tableau[i][a];
+            }
+            return 3U;
+        }
+        case qc::OpType::Z: { // Z = S x S
+            if (gate->isControlled()) {
+                util::fatal("Expected single-qubit gate");
+            }
+            const auto a = gate->getTargets().at(0U);
+            for (auto i = 0U; i < nqubits; i++) {
+                tableau[i][2U * nqubits] ^= tableau[i][a] & tableau[i][a + nqubits];
+                tableau[i][a + nqubits] ^= tableau[i][a];
+            }
+            for (auto i = 0U; i < nqubits; i++) {
+                tableau[i][2U * nqubits] ^= tableau[i][a] & tableau[i][a + nqubits];
+                tableau[i][a + nqubits] ^= tableau[i][a];
+            }
+            return 2U;
+        }
+        case qc::OpType::Y: { // Y = H x S x S x H x S x S
+            if (gate->isControlled()) {
+                util::fatal("Expected single-qubit gate");
+            }
+            const auto a = gate->getTargets().at(0U);
+            for (auto i = 0U; i < nqubits; i++) {
+                tableau[i][2U * nqubits] ^= tableau[i][a] & tableau[i][a + nqubits];
+                std::swap(tableau[i][a], tableau[i][a + nqubits]);
+            }
+            for (auto i = 0U; i < nqubits; i++) {
+                tableau[i][2U * nqubits] ^= tableau[i][a] & tableau[i][a + nqubits];
+                tableau[i][a + nqubits] ^= tableau[i][a];
+            }
+            for (auto i = 0U; i < nqubits; i++) {
+                tableau[i][2U * nqubits] ^= tableau[i][a] & tableau[i][a + nqubits];
+                tableau[i][a + nqubits] ^= tableau[i][a];
+            }
+            for (auto i = 0U; i < nqubits; i++) {
+                tableau[i][2U * nqubits] ^= tableau[i][a] & tableau[i][a + nqubits];
+                std::swap(tableau[i][a], tableau[i][a + nqubits]);
+            }
+            for (auto i = 0U; i < nqubits; i++) {
+                tableau[i][2U * nqubits] ^= tableau[i][a] & tableau[i][a + nqubits];
+                tableau[i][a + nqubits] ^= tableau[i][a];
+            }
+            for (auto i = 0U; i < nqubits; i++) {
+                tableau[i][2U * nqubits] ^= tableau[i][a] & tableau[i][a + nqubits];
+                tableau[i][a + nqubits] ^= tableau[i][a];
+            }
+            return 6U;
+        }
+        default:
+            util::fatal("Unsupported gate encountered: " + std::to_string(gate->getType()));
+            break;
+    }
+    return 0U;
+}
+
+Tableau Tableau::getDiagonalTableau(int nQubits) {
+    TableauType result{};
+    result.resize(nQubits);
+    for (auto i = 0; i < nQubits; i++) {
+        result[i].resize(2U * nQubits + 1U);
+        for (auto j = 0; j < 2 * nQubits; j++) {
+            if (i == j - nQubits) {
+                result[i][j] = 1;
+            } else {
+                result[i][j] = 0;
+            }
```

Comment:
this else branch can be dropped since the default initialisation is `0` anyways.